src/Server/RemoteFSHandler.cpp

```diff
@@ -0,0 +1,543 @@
+            disk->removeRecursive(path);
+            writeVarUInt(RemoteFSProtocol::RemoveRecursive, *out);
+            out->next();
+            break;
+        case RemoteFSProtocol::SetLastModified:
+            receivePath(path);
+            readVarUInt(time_var, *in);
+            disk->setLastModified(path, Poco::Timestamp::fromEpochTime(time_var));
+            writeVarUInt(RemoteFSProtocol::SetLastModified, *out);
+            out->next();
+            break;
+        case RemoteFSProtocol::GetLastModified:
+            receivePath(path);
+            time_var = disk->getLastModified(path).epochTime();
+            writeVarUInt(RemoteFSProtocol::GetLastModified, *out);
+            writeVarUInt(time_var, *out);
+            out->next();
+            break;
+        case RemoteFSProtocol::GetLastChanged:
+            receivePath(path);
+            time_var = disk->getLastChanged(path);
+            writeVarUInt(RemoteFSProtocol::GetLastChanged, *out);
+            writeVarUInt(time_var, *out);
+            out->next();
+            break;
+        case RemoteFSProtocol::SetReadOnly:
+            receivePath(path);
+            disk->setReadOnly(path);
+            writeVarUInt(RemoteFSProtocol::SetReadOnly, *out);
+            out->next();
+            break;
+        case RemoteFSProtocol::CreateHardLink:
+            receivePath(path);
+            receivePath(path2);
+            disk->createHardLink(path, path2);
+            writeVarUInt(RemoteFSProtocol::CreateHardLink, *out);
+            out->next();
+            break;
+        case RemoteFSProtocol::TruncateFile:
+            receivePath(path);
+            readVarUInt(size_t_var, *in);
+            disk->truncateFile(path, size_t_var);
+            writeVarUInt(RemoteFSProtocol::TruncateFile, *out);
+            out->next();
+            break;
+        default:
+            throw Exception(ErrorCodes::UNKNOWN_PACKET_FROM_CLIENT, "Unknown packet {} from client", toString(packet_type));
+    }
+}
+
+void RemoteFSHandler::receivePath(std::string & path)
+{
+    readStringBinary(path, *in);
+    LOG_TRACE(log, "Received path {}", path);
+}
+
+void RemoteFSHandler::receiveUnexpectedHello()
+{
+    String skip_string;
+
+    readStringBinary(skip_string, *in);
+
+    throw NetException(ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT, "Unexpected packet Hello received from client");
+}
+
+void RemoteFSHandler::iterateDirectory()
+{
+    std::string path;
+    receivePath(path);
+    for (auto iter = disk->iterateDirectory(path); iter->isValid(); iter->next())
+    {
+        LOG_TRACE(log, "Writing dir entry {}", iter->path());
+        writeVarUInt(RemoteFSProtocol::DataPacket, *out);
+        writeStringBinary(iter->path(), *out);
+    }
+    writeVarUInt(RemoteFSProtocol::EndIterateDirectory, *out);
+    out->next();
+}
+
+void RemoteFSHandler::listFiles()
+{
+    std::string path;
+    receivePath(path);
+    std::vector<String> files;
+    disk->listFiles(path, files);
```

Comment:
better to use `iterateDirectory` method, it will be more efficient and we avoid getting all files in memory (usually `listFiles` uses `iterateDirectory` inside)